### ISSUE TITLE
Support for Blizzard Interactives new Dev/OAuth API

### ISSUE THREAD

**Issue by madmagi** [RockstarLabs/OwinOAuthProviders#28]
Not really an issue as much as a request.

Blizzard Interactive has announced a new set of API's for their systems, including publicly available ones that are trimmed with OAuth security.  

https://dev.battle.net

I tried to use the 'generic' OAuth objects in the Owin.Security.Providers to no avail.  Any chance you could take a look at their implementation and see what it would take to support it?

I have an active blizzard account and so would be able to help as best I can.

Thank You

**Comment by jerriep** [comment]
@madmagi My apologies for not getting back to you sooner. Unfortunately I do not have the capacity right now to do this. Hopefully someone else is willing to pick it up and do it, otherwise you may have to wait a while. Sorry :(

**Comment by KetaSwe** [comment]
I have worked a bit with this and I think I have a working version of the Battle.net OAuth Authentication.
I have never commited anything to Github before so I don't know if I'm able to add it to this Github repo.

So what should I do? I probably need to learn Github at some point. Maybe I could start my own repo and add it there just to learn how Github works and so I don't mess up this repo.

**Comment by jerriep** [comment]
Thank you @KetaSwe . You will need to submit a pull request. Please see the following link:
https://help.github.com/categories/collaborating/

Read the first 2 sections called "Using pull requests" and "Creating a pull request".

Shout if you get stuck

**Comment by KetaSwe** [comment]
Thanks!

I will try to do it as soon as possible but it might take a few days/weeks depending on how much spare time I have.

**Comment by jerriep** [comment]
No problem :) Thanks

**Comment by KetaSwe** [comment]
There, I've done a pull request. I hope that I haven't missed anything.

**Comment by jerriep** [comment]
Will check it out sometime over the weekend, thanks :)

**Comment by ericwj** [comment]
This code is hardcoded to work only against the authentication servers in Europe. I'd like this to be generalized - for starters by including more of the configuration strings in the options class as I see it is done for example in the vNext code. I'll give it a try if you're OK with that. 

Are there plans to port this package to ASP.NET 5? Or is there another repository for that? Seeing the code for the Microsoft packages for Microsoft accounts, Google, Facebook and Twitter in https://github.com/aspnet/Security/ that shouldn't be hard to do at all.

**Comment by KetaSwe** [comment]
Ah, you are right. My mistake and I'll look into it as soon as possible.

**Comment by jerriep** [comment]
@ericwj As far as ASP.NET 5 is concerned, I reached out to see whether I could upgrade these providers to support ASP.NET 5 and then integrate it into the aspnet/Security repo, but they were not keen on that and suggested that they may set up another organization where "community" projects like these go.

That being said, once ASP.NET 5 comes out, the need for this project will go away almost completely. They have added a new, very generic, OAuth provider which can be make to work against most services with a little bit of configuration. See https://github.com/aspnet/Security/blob/dev/samples/SocialSample/Startup.cs

Sometimes the configuration can become a little bit elaborate, as you can see in some of those demos, so what I am in the process of doing is setting up a website (www.oauthforaspnet.com) where I will give people the correct source code to configure ASP.NET 5 to work against the service of their choice.

I have not added this work to the live website yet, as the stuff around the Security providers is too much in flux at the moment. Maybe by the time the next beta hits it will be at such a stage that I can start doing my work to add them to the website.

**Comment by ericwj** [comment]
Yeah if it's in that repository, it actually means Microsoft will actively support it even by phone and whatnot, so they will want to keep it fairly limited.

Yet to only have code samples to copy from a website is perhaps a step back. I can imagine it'd be more valuable even to have each provider have its own .nupkg in the feed - that would fit nicely with the strategy for NuGet going forward.

Those projects will be mostly very small as a result of the generic one being - well, generic, but then there's special stuff that can go in such a project as well, for example the 5 OAuth servers that Blizzard actually has and each provider has quirks, special claims to make or data that would be useful to type just once...the ideal would still be to only have to add a package, type a secret or two and get going, shouldn't it be?

**Comment by ericwj** [comment]
@KetaSwe didn't mean to put the heat on you, I wrote some code myself already ;)

**Comment by jerriep** [comment]
@ericwj Thanks for your feedback. Once the Security API stabilizes a bit more and I have a clearer direction of exactly what Microsoft is doing with the new versions I will re-evaluate. If it makes sense to write tiny little wrappers arounds the generic one in certain instances then we can certainly approach it like that.

**Comment by KetaSwe** [comment]
I have made a pull request now with changes to the battle.net provider. I added a Region property in Options where you can specify the region you want for the authentication.

**Comment by akoltz** [comment]
What's the expected pattern if you want to authenticate against any region?  Can you register BattleNetAuthenticationOptions multiple times with different Regions?

**Comment by jerriep** [comment]
I am not sure about the Blizzard implementation, but here is a blog post about registering providers multiple times. Maybe it will help:

http://umbraco.com/follow-us/blog-archive/2015/3/26/configuring-aspnet-identity-oauth-login-providers-for-multi-tenancy
